### PR TITLE
Minor refactor

### PR DESCRIPTION
**Pull request by satya164**

### CODE REVIEW

File: src/PaletteProvider.js [hunk 1]
```diff
@@ -54,42 +50,28 @@ type Props = {
   /**
    * Render LoaderComponent when the palette is being created
    */
-  LoaderComponent?:
-    | React$Component<*, *, *>
-    | ((...args: *) => React$Element<*>),
+  LoaderComponent?: ReactClass<*> | ((...args: *) => React.Element<*>),
```
Comment (rgommezz) [inline src/PaletteProvider.js]: Let's use `Array<*>` or `Array<any>` better for the rest parameter.

From the docs: `If you add a type annotation to a rest parameter, it must always explicitly be an Array type.`

https://flow.org/en/docs/types/functions/#toc-rest-parameters

File: src/PaletteProvider.js [hunk 2]
```diff
@@ -144,35 +126,47 @@ export default class MaterialPaletteProvider
...
         const paletteWithDefaults = this._mergeWithDefaults(palette);
-        execIfFunction(this.props.onFinish, paletteWithDefaults);
-        if (!this.props.forceRender) {
-          this.setState({ palette: paletteWithDefaults });
-        }
+
+        if (onFinish) onFinish(paletteWithDefaults);
+
+        this.setState({ palette: paletteWithDefaults });
         this.eventEmitter.publish({
```
Comment (rgommezz) [inline src/PaletteProvider.js]: We need to `publish` before setting state, remember the issue we discussed at the office. The change is in my branch anyway, but will be good to have it here as well already 🤖

File: src/PaletteProvider.js [hunk 3]
```diff
@@ -144,35 +126,47 @@ export default class MaterialPaletteProvider
...
-    } else if (!this.state.palette && !this.props.forceRender) {
-      return null;
+    const { forceRender, LoaderComponent, children } = this.props;
+    const shouldRender = this.state.palette || forceRender;
+
+    if (LoaderComponent && !this.state.palette) {
+      return <LoaderComponent />;
+    }
+
+    if (shouldRender) {
```
Comment (rgommezz) [inline src/PaletteProvider.js:166]: shouldn't we change the order? If `forceRender` is provided, it seems natural for me to override `LoaderComponent`, what do you think?

File: src/__tests__/PaletteProvider.test.js [hunk 4]
```diff
@@ -102,33 +102,24 @@ describe('PaletteProvider', () => {
...
-          <Text>Test</Text>
-        </PaletteProvider>,
-      );
+  it('should throw error if `onError` was not passed and palette creation fails', () => {
+    createMaterialPalette.mockImplementation(() => ({
+      then(success, failure) {
+        try {
+          failure(new Error('test'));
+        } catch (error) {
+          expect(error.message).toMatch(/MaterialPaletteProvider.*test/);
```
Comment (rgommezz) [inline src/__tests__/PaletteProvider.test.js]: Why not `expect(error.message).toBe('Uncaught MaterialPaletteProvider exception: test')` for more accuracy of what's being thrown?

File: src/__tests__/withPalette.test.js [hunk 5]
```diff
@@ -159,19 +160,17 @@ describe('withPalette', () => {
       expect(palette).toEqual({
         lightVibrant: { ...localDefaults.lightVibrant, population: 0 },
       });
-      expect(style).toEqual([{ fontSize: '14px' }, {}]);
+      expect(StyleSheet.flatten(style)).toEqual({ fontSize: '14px' });
```
Comment (rgommezz) [inline src/__tests__/withPalette.test.js:163]: Nice! Didn't know about `flatten` 👍

**Review by rgommezz — CHANGES_REQUESTED** [review]

**Comment by rgommezz** [comment]
Also, we need to update the API docs to remove `onInit` from there

File: src/__tests__/PaletteProvider.test.js (hunk 4 above)
Comment (satya164) [reply src/__tests__/PaletteProvider.test.js]: Blame @zamotany xD

File: src/PaletteProvider.js (hunk 3 above)
Comment (satya164) [reply src/PaletteProvider.js:166]: Just following what's in the comment in the prop types. cc @zamotany

Comment (zamotany) [reply src/PaletteProvider.js:166]: From the docs: _Does not take effect if LoaderComponent is specified!_

Besides, why would you specify `LoaderComponent` when you would force the render anyway?

**Review by zamotany — COMMENTED** [review]
Why are ~we~ you removing `onInit`??

**Comment by satya164** [comment]
@zamotany it's basically the same as when the provider mounts, so seems unnecessary

**Review by rgommezz — APPROVED** [review]

**Pull request merged**